rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md

```diff
@@ -0,0 +1,229 @@
+
+3. **MinCapacity Worker Stability** - Calculate percentage of time pools maintain `minCapacity` without deficits
+
+**Secondary Metrics:**
+
+4. **Cache Preservation Effectiveness** - Compare 90th percentile worker uptime between `minCapacity` and overflow workers to measure cache retention
+
+5. **Termination Frequency Impact** - Monitor rate of `minCapacity` worker terminations per hour
+
+6. **Capacity Response Time** - Average time to fulfill `minCapacity` requirements after capacity deficits occur
+
+**Success Criteria:**
+- **Task wait time**: 20% reduction in P95 pending→running time
+- **Worker churn**: 50% reduction in worker spawn frequency for `minCapacity` pools
+- **Cache effectiveness**: 80% of `minCapacity` workers have >1 hour uptime
+- **Deficit resolution**: <60 seconds average time to resolve `minCapacity` deficits
+- **Termination cost**: <5% of `minCapacity` workers terminated for capacity reduction per day
+
+**Trade-off Evaluation:**
+Cost-benefit analysis will compare the benefits of reduced task wait times (measured by time savings multiplied by task throughput) against the costs of cache loss from forced terminations (measured by termination rate multiplied by average cache warmup time).
+
+## Rollout Strategy
+
+**Phase 1**: Single Pool Testing
+- Enable on one stable test pool (e.g., `gecko-1/decision-gcp` for try branch)
+- Monitor success metrics, cache preservation, and termination frequency
+- Validate that pool benefits outweigh costs of occasional forceful termination
+
+**Phase 2**: Gradual Expansion
+- Roll out to remaining decision pools with stable `minCapacity` requirements
+- Prioritize pools where `minCapacity` changes are infrequent
+- Monitor trade-off between cache benefits and capacity management responsiveness
+
+**Phase 3**: Full Deployment
+- Enable for all eligible pools after validating net positive impact
+- Continue monitoring optimization effectiveness across different usage patterns
+- Consider disabling for pools where `minCapacity` changes are too frequent
+
+## Error Handling and Edge Cases
+
+**Worker Lifecycle Management:**
+- **Pool reconfiguration**: Capacity changes trigger worker replacement, not reconfiguration
+- **Graceful transitions**: When possible, only terminate idle workers to preserve active caches
+- **Resource allocation**: `minCapacity` workers mixed with overflow workers on same infrastructure
+
+**Capacity Reduction Strategy:**
+When `minCapacity` decreases, a hybrid approach minimizes cache loss by first prioritizing termination of idle excess workers to preserve active caches. If additional capacity reduction is needed, busy workers are terminated immediately, starting with the oldest workers first to maximize cache preservation for recently started workers.
+
+## Compatibility Considerations
+
+**Backward Compatibility:**
+- Opt-in via `minCapacityIdleTimeoutSecs` configuration
+- Existing pools continue current behavior unless explicitly enabled
```

Comment:
Honestly, I'd consider implementing this would be a "bug fix" to worker manager and not a new feature that we'd need to have backward compatibility for to the old, "broken" `minCapacity` functionality.

Reply:
Now that I think about this more, I think this is fine, so long as we properly address changes to the `minCapacity` pool config. We could err on the side of caution and put it behind a feature flag at first to only enable on certain pools for testing/validating, then later drop it, and make that behavior the default for `minCapacity`.

Reply:
Yeah, I believe we should make the behavior the default one once we're confident enough with the change. So, I'd be fine to `minCapacityIdleTimeoutSecs` whenever we don't need it anymore.